core/src/main/java/org/infinispan/functional/impl/SimpleReadWriteMapImpl.java

```diff
@@ -0,0 +1,195 @@
+
+         results.add(CompletionStages.join(cf));
+      }
+
+      return Traversables.of(results.stream());
+   }
+
+   @Override
+   public <R> Traversable<R> evalMany(Set<? extends K> keys, Function<EntryView.ReadWriteEntryView<K, V>, R> f) {
+      List<R> results = new ArrayList<>(keys.size());
+      BiFunction<?, EntryView.ReadWriteEntryView<K, V>, R> bf = (ignore, v) -> f.apply(v);
+      for (K key : keys) {
+         CompletableFuture<R> cf = eval(key, null, bf);
+         assert cf.isDone() : "eval() returned an incomplete CompletableFuture";
+
+         results.add(CompletionStages.join(cf));
+      }
+
+      return Traversables.of(results.stream());
+   }
+
+   @Override
+   public <R> Traversable<R> evalAll(Function<EntryView.ReadWriteEntryView<K, V>, R> f) {
+      return evalMany(fmap.cache.keySet(), f);
+   }
+
+   private boolean isStatisticsEnabled() {
+      return !params.containsAll(Param.StatisticsMode.SKIP);
+   }
+
+   private boolean hasListeners() {
+      return !notifier.getListeners().isEmpty();
+   }
+
+   private boolean canHandleEntry(InternalCacheEntry<K, V> ice) {
+      if (ice == null) {
+         return true;
+      }
+
+      if (ice.canExpire()) {
+         if (ice.isExpired(timeService.wallClockTime())) {
+            if (notifier.hasListener(CacheEntryExpired.class)) {
+               CompletionStages.join(notifier.notifyCacheEntryExpired(ice.getKey(), ice.getValue(),
+                     ice.getMetadata(), ImmutableContext.INSTANCE));
+            }
+         }
+         return false;
+      }
+      return true;
+   }
+
+   private <R, T> R evalInternal(K key, T argument, BiFunction<T, EntryView.ReadWriteEntryView<K, V>, R> f) {
+      Objects.requireNonNull(key, KEY_CANNOT_BE_NULL);
+      Objects.requireNonNull(f, FUNCTION_CANNOT_BE_NULL);
+      ByRef<R> result = new ByRef<>(null);
+      fmap.cache.getDataContainer().compute(key, (k, oldEntry, factory) -> compute(key, argument, f, oldEntry, factory, result));
+      return result.get();
+   }
+
+   private <R, T> InternalCacheEntry<K, V> compute(K key, T argument, BiFunction<T, EntryView.ReadWriteEntryView<K, V>, R> f,
+                                                   InternalCacheEntry<K, V> oldEntry, InternalEntryFactory factory,
+                                                   ByRef<R> holder) {
+      MVCCEntry<K, V> e = extractEntry(oldEntry, key);
+      EntryViews.AccessLoggingReadWriteView<K, V> view =  EntryViews.readWrite(e, fmap.cache.getKeyDataConversion(), fmap.cache.getValueDataConversion());
+      holder.set(EntryViews.snapshot(f.apply(argument, view)));
+      if (hasListeners()) {
+         if (oldEntry == null) {
+            CompletionStages.join(notifier.notifyCacheEntryCreated(key, e.getValue(), e.getOldMetadata(), true, ImmutableContext.INSTANCE, null));
+            CompletionStages.join(notifier.notifyCacheEntryCreated(key, e.getValue(), e.getOldMetadata(), false, ImmutableContext.INSTANCE, null));
```

Comment:
We should fire these outside of the CHM lock if we want to behave the same as the SimpleCacheImpl methods.

Reply:
Updated. The pre notification fires from within the compute and the pos from outside.